cgp/genome.py

```diff
@@ -354,79 +376,83 @@ def mutate(self, mutation_rate: float, rng: np.random.RandomState):
         mutation_rate : float
             Proportion of genes to be mutated, between 0 and 1.
         rng : numpy.random.RandomState
-            Random number generator instance to use for crossover.
+            Random number generator instance to use for mutation.
 
         Returns
         ----------
         bool
             True if only inactive regions of the genome were mutated, False otherwise.
         """
 
-        def count_dna_differences(dna0: List[int], dna1: List[int]) -> int:
-            return len([1 for gene0, gene1 in zip(dna0, dna1) if gene0 != gene1])
-
-        n_mutations = int(mutation_rate * len(self.dna))
-        assert n_mutations > 0
-
         graph = CartesianGraph(self)
         active_regions = graph.determine_active_regions()
-
         dna = list(self._dna)
-
         only_silent_mutations = True
-        while count_dna_differences(self.dna, dna) < n_mutations:
 
-            gene_idx = rng.randint(0, self._n_genes)
-            region_idx = gene_idx // self._length_per_region
+        selected_gene_indices = self._select_gene_indices_for_mutation(mutation_rate, len(dna))
 
-            if self._is_input_region(region_idx):
-                continue  # nothing to do here
+        for (gene_idx, allele) in zip(selected_gene_indices, np.array(dna)[selected_gene_indices]):
 
-            elif self._is_output_region(region_idx):
-                silent = self._mutate_output_region(dna, gene_idx, rng)
+            region_idx = gene_idx // self._length_per_region
 
-            elif self._is_hidden_region(region_idx):
-                silent = self._mutate_hidden_region(dna, gene_idx, active_regions, rng)
+            permissible_values = self._determine_alternative_permissible_values(
```

Comment:
One question: Can this check be done once at initialization time of the `Genome` such that we don't need to repeat this every time we want to mutate the genome? The permissible values only depend on the position of the gene in the dna, right? And since the genome parameters don't change over time, I am suspecting that it might be possible to do this once and be done with it. This might save some computation time.

Reply:
in principle yes, i think. one could build a dictionary, e.g., in `__init__`  that lists all permissible values for each gene idx. i agree, this might save quite some computation time at the expense of increased memory usage and slower instantiation of `Genome` objects. as always, the root of all evil is premature optimization, so i wouldn't implement this without profiling and finding that this is indeed a bottleneck. maybe a more lightweight compromise would be some kind of memoization, e.g., decorating the function determining all permissible values with the [lru cache decorator](https://docs.python.org/3/library/functools.html#functools.lru_cache). we should address this, and in general some lightweight performance improvements, in a separate issue/PRs.

Reply:
Thanks for the calendar motto ;-)
I think that the advantages of performing this calculation instead of many times outweigh the disadvantage of increased memory consumption (This will be a tiny data structure of a few bytes). But I am fine with postponing this to another PR. I created a ticket to remind us: #196

Reply:
That's a good input, thanks! Yes I don't see why this shouldn't work. I agree that memory shouldn't be an issue, the number of values to store is `n < n_units * len(dna)`. I will take care of the #196 once I made this PR ready to merge